### ISSUE TITLE
Definition of faithfulness for Text-NMN

### ISSUE THREAD

**Issue by ramild** [allenai/faithful-nmn#1]
Hi!
I would like to thank the authors of the paper for the amazing work that they did, also open-sourcing all the necessary code and data. It is a great contribution to the community!

I am very interested in reproducing the results for Text-NMN. I managed to get similar accuracy for the Text-NMN as reported in the paper. But in the process of studying the code, I faced an important question that I couldn't figure out by myself. I would appreciate it very much if the authors could help me out with them!

Broadly, I couldn't manage to understand how exactly the faithfulness score is being measured. As far, as I can see INTERPRET=true mode is being used according to https://github.com/nitishgupta/nmn-drop/blob/interpret/scripts/allennlp/models/drop_bert/predict_interpret.sh#L8 when measuring the faithfulness. Then from the lines https://github.com/nitishgupta/nmn-drop/blob/interpret/semqa/models/drop_parser_bert.py#L616 , it follows that in this mode the model is allowed to see the gold action sequences in the process of the constrained beam search. As far as I understand, it means that the model doesn't rely on the trained parser and instead has access to the gold programs. Is it indeed the case? Doesn't it make more sense to perform an unconstrained beam search, similarly to how it should work at prediction time (https://github.com/nitishgupta/nmn-drop/blob/interpret/semqa/models/drop_parser_bert.py#L600 )? Isn't the goal to compute faithfulness at prediction time or did I misunderstand this part?

Additionally, I noticed that when the faithfulness score is computed, the module outputs from gold and prediction are being compared even if the two programs are different. But why would we want to compute cross-entropy between outputs of the two modules of different functionality (like "find" and "find-num")? Especially when the lengths of the program sequences are different. Wouldn't it be more natural to compute the faithfulness only for the prediction programs that match the gold ones, and for others define the faithfulness score as zero (since, if the program isn't correct, it isn't faithful)?

**Comment by benbogin** [comment]
Hi Ramil, thanks for your interest! @nitishgupta has worked on this part - Nitish can you see if you can help?

**Comment by ramild** [comment]
@nitishgupta Hi, I would be enormously greatful if you could give a short reply by confirming/refuting my current understanding!
Working on this line of research for DROP, I already spent many hours exploring the code, but got stuck a bit, and your answer would be extremely helpful.

Thank you very much for your work.

**Comment by nitishgupta** [comment]
Hi Ramil, 

Extremely sorry for the delay in response. I was in the processing of moving to a different country and completely lost track of emails. 

1. You are correct that we are computing the faithfulness score for the execution of the GOLD program (in Text-NMN this is attention over paragraph tokens) during prediction time. Do note that the output of the neural modules (e.g. find) is being predicted by the learned modules. We compute this by executing the gold program since the idea is --- even if the parser was perfect and predicted the correct program, is the execution that is learned by the neural modules faithful to their intended behavior. Btw, this is also true for the Visual-NMN experiments. 

2. `when the faithfulness score is computed, the module outputs from gold and prediction are being compared even if the two programs are different.` - Could you please refer to when this is happening. My guess is you are referring to the part of the code that corresponds to one of the ablation studies we did. Where the idea was to see if certain modules (e.g. find-max-num) didn't exist in the NMN repo, how would that affect the learning of other modules. In that case, we compare outputs for modules where the programs are different, but the comparison happens for the same module. For example, comparing the outputs of `find` in `find[]` and `find-max-num(find[])`.

Please let me know if those answer your questions.
